File: macos/Tests/VPSQuotaCoreTests/SchemaConsistencyTests.swift
import Foundation
import SQLite3
import Testing
@testable import VPSQuotaCore

/// `shared/schema.sql` 声称是两端建表语句的唯一事实源，但三份拷贝靠人工同步迟早会漂移。
/// 这组测试直接建一个真库，把它实际的表结构和 schema.sql 声明的对照，把约定变成保证。
///
/// Windows 端的 SqliteStore.cs 用的是同一段 SQL 文本，这里过了也就等于那边过了。
@Suite("Schema 一致性")
struct SchemaConsistencyTests {

    /// 从 schema.sql 里解析出「表名 → [(列名, 类型)]」，忽略注释与空白差异。
    private func declaredTables() throws -> [String: [(String, String)]] {
        let url = URL(fileURLWithPath: #filePath)
            .deletingLastPathComponent()
            .deletingLastPathComponent()
            .deletingLastPathComponent()
            .deletingLastPathComponent()
            .appendingPathComponent("shared/schema.sql")
        var sql = try String(contentsOf: url, encoding: .utf8)

        // 去掉 -- 行注释，它们只影响可读性，不影响结构
        sql = sql.split(separator: "\n", omittingEmptySubsequences: false)
            .map { line -> String in
                guard let range = line.range(of: "--") else { return String(line) }
                return String(line[..<range.lowerBound])
            }
            .joined(separator: "\n")

        var result: [String: [(String, String)]] = [:]
        let pattern = #"CREATE TABLE IF NOT EXISTS\s+(\w+)\s*\(([^;]*)\)\s*;"#
        let regex = try NSRegularExpression(pattern: pattern, options: [.dotMatchesLineSeparators])
        let range = NSRange(sql.startIndex..., in: sql)

        for match in regex.matches(in: sql, range: range) {
            guard let nameRange = Range(match.range(at: 1), in: sql),
                  let bodyRange = Range(match.range(at: 2), in: sql) else { continue }
            let table = String(sql[nameRange])

            var columns: [(String, String)] = []
            for part in sql[bodyRange].split(separator: ",") {
                let tokens = part.split(whereSeparator: { $0 == " " || $0 == "\n" || $0 == "\t" })
                // 跳过 PRIMARY KEY(...) 这类表级约束
                guard let first = tokens.first, first.uppercased() != "PRIMARY",
                      tokens.count >= 2 else { continue }
                columns.append((String(first), String(tokens[1]).uppercased()))
            }
            result[table] = columns
        }
        return result
    }

    /// 读真实数据库里的表结构。
    private func actualColumns(db: OpaquePointer, table: String) -> [(String, String)] {
        var stmt: OpaquePointer?
        guard sqlite3_prepare_v2(db, "PRAGMA table_info(\(table))", -1, &stmt, nil) == SQLITE_OK
        else { return [] }
        defer { sqlite3_finalize(stmt) }

        var result: [(String, String)] = []
        while sqlite3_step(stmt) == SQLITE_ROW {
            guard let name = sqlite3_column_text(stmt, 1),
                  let type = sqlite3_column_text(stmt, 2) else { continue }
            result.append((String(cString: name), String(cString: type).uppercased()))
        }
        return result
    }

    @Test("实际建出来的表结构与 shared/schema.sql 逐列一致")
    func storeMatchesSharedSchema() async throws {
        let declared = try declaredTables()
        #expect(declared.keys.sorted() == ["daily_usage", "fetch_log"])

        let dir = URL(fileURLWithPath: NSTemporaryDirectory())
            .appendingPathComponent("vpsquota-schema-\(UUID().uuidString)", isDirectory: true)
        defer { try? FileManager.default.removeItem(at: dir) }
        let file = dir.appendingPathComponent("usage.sqlite")

        // 先让 SQLiteStore 按它自己的 migrate() 建库
        _ = try SQLiteStore(path: file)

        var db: OpaquePointer?
        #expect(sqlite3_open(file.path, &db) == SQLITE_OK)
        defer { sqlite3_close(db) }
        let handle = try #require(db)

        for (table, expected) in declared {
            let actual = actualColumns(db: handle, table: table)
            #expect(
                actual.map(\.0) == expected.map(\.0),
                "表 \(table) 的列名与 shared/schema.sql 不一致"
            )
            #expect(
                actual.map(\.1) == expected.map(\.1),
                "表 \(table) 的列类型与 shared/schema.sql 不一致"
            )
        }
    }

    @Test("索引也一并建出来了")
    func indexesExist() async throws {
        let dir = URL(fileURLWithPath: NSTemporaryDirectory())
            .appendingPathComponent("vpsquota-index-\(UUID().uuidString)", isDirectory: true)
        defer { try? FileManager.default.removeItem(at: dir) }
        let file = dir.appendingPathComponent("usage.sqlite")
        _ = try SQLiteStore(path: file)

        var db: OpaquePointer?
        #expect(sqlite3_open(file.path, &db) == SQLITE_OK)
        defer { sqlite3_close(db) }

        var stmt: OpaquePointer?
        #expect(sqlite3_prepare_v2(
            db, "SELECT name FROM sqlite_master WHERE type = 'index' ORDER BY name", -1, &stmt, nil
        ) == SQLITE_OK)
        defer { sqlite3_finalize(stmt) }

        var names: [String] = []
        while sqlite3_step(stmt) == SQLITE_ROW {
            if let n = sqlite3_column_text(stmt, 0) { names.append(String(cString: n)) }
        }
        #expect(names.contains("idx_daily_usage_day"))
        #expect(names.contains("idx_fetch_log_server"))
    }
}

File: windows/VpsQuota.Tests/SchemaConsistencyTests.cs
namespace VpsQuota.Tests;

using System.IO;
using System.Runtime.CompilerServices;
using System.Text.RegularExpressions;
using Microsoft.Data.Sqlite;
using VpsQuota.Storage;
using Xunit;

/// <summary>
/// <c>shared/schema.sql</c> 声称是两端建表语句的唯一事实源，但三份拷贝靠人工同步迟早会漂移。
/// </summary>
/// <remarks>
/// macOS 端有同名的一组测试，但它只验证 Swift 那一份拷贝 —— 一行 C# 都不碰。
/// 换句话说，此前没有任何机制阻止 <see cref="SqliteStore"/> 里那段 SQL 独自漂移。
/// 这组测试建一个真库，把它<b>实际</b>的表结构与 schema.sql 声明的对照。
/// </remarks>
public class SchemaConsistencyTests
{
    private static string RepoRoot([CallerFilePath] string? thisFile = null)
    {
        // windows/VpsQuota.Tests/SchemaConsistencyTests.cs → 上溯三级即仓库根
        var directory = Path.GetDirectoryName(thisFile)!;
        return Path.GetFullPath(Path.Combine(directory, "..", ".."));
    }

    /// <summary>从 schema.sql 解析出「表名 → [(列名, 类型)]」，忽略注释与空白差异。</summary>
    private static Dictionary<string, List<(string Name, string Type)>> DeclaredTables()
    {
        var sql = File.ReadAllText(Path.Combine(RepoRoot(), "shared", "schema.sql"));

        // 去掉 -- 行注释，它们只影响可读性，不影响结构
        sql = string.Join("\n", sql.Split('\n').Select(line =>
        {
            var index = line.IndexOf("--", StringComparison.Ordinal);
            return index < 0 ? line : line[..index];
        }));

        var result = new Dictionary<string, List<(string, string)>>();
        var matches = Regex.Matches(
            sql, @"CREATE TABLE IF NOT EXISTS\s+(\w+)\s*\(([^;]*)\)\s*;", RegexOptions.Singleline);

        foreach (Match match in matches)
        {
            var columns = new List<(string, string)>();
            foreach (var part in match.Groups[2].Value.Split(','))
            {
                var tokens = part.Split(
                    new[] { ' ', '\n', '\t', '\r' }, StringSplitOptions.RemoveEmptyEntries);
                // 跳过 PRIMARY KEY(...) 这类表级约束
                if (tokens.Length < 2 || tokens[0].Equals("PRIMARY", StringComparison.OrdinalIgnoreCase))
                    continue;
                columns.Add((tokens[0], tokens[1].ToUpperInvariant()));
            }
            result[match.Groups[1].Value] = columns;
        }

        return result;
    }

    /// <summary>建一个真库，读回它实际的表结构。</summary>
    private static async Task<Dictionary<string, List<(string Name, string Type)>>> ActualTablesAsync()
    {
        var directory = Path.Combine(Path.GetTempPath(), "vpsquota-schema-" + Guid.NewGuid().ToString("n"));
        var path = Path.Combine(directory, "usage.sqlite");

        await using (var store = new SqliteStore(path)) { /* 建库即可 */ }

        var result = new Dictionary<string, List<(string, string)>>();
        using (var connection = new SqliteConnection(new SqliteConnectionStringBuilder
        {
            DataSource = path,
            Mode = SqliteOpenMode.ReadOnly,
        }.ToString()))
        {
            connection.Open();
            foreach (var table in new[] { "daily_usage", "fetch_log", "server_meta" })
            {
                var columns = new List<(string, string)>();
                using var command = connection.CreateCommand();
                command.CommandText = $"PRAGMA table_info({table})";
                using var reader = command.ExecuteReader();
                while (reader.Read())
                {
                    columns.Add((reader.GetString(1), reader.GetString(2).ToUpperInvariant()));
                }
                result[table] = columns;
            }
        }

        try { Directory.Delete(directory, recursive: true); } catch (IOException) { }
        return result;
    }

    [Fact(DisplayName = "实际建出来的表结构与 shared/schema.sql 逐列一致")]
    public async Task TablesMatchSharedSchema()
    {
        var declared = DeclaredTables();
        var actual = await ActualTablesAsync();

        Assert.Equal(3, declared.Count);   // daily_usage + fetch_log + server_meta

        foreach (var (table, columns) in declared)
        {
            Assert.True(actual.ContainsKey(table), $"代码没有建出表 {table}");
            Assert.Equal(columns, actual[table]);
        }
    }

    [Fact(DisplayName = "索引也一并建出来了")]
    public async Task IndexesExist()
    {
        var directory = Path.Combine(Path.GetTempPath(), "vpsquota-index-" + Guid.NewGuid().ToString("n"));
        var path = Path.Combine(directory, "usage.sqlite");

        await using (var store = new SqliteStore(path)) { }

        var names = new List<string>();
        using (var connection = new SqliteConnection(new SqliteConnectionStringBuilder
        {
            DataSource = path,
            Mode = SqliteOpenMode.ReadOnly,
        }.ToString()))
        {
            connection.Open();
            using var command = connection.CreateCommand();
            command.CommandText = "SELECT name FROM sqlite_master WHERE type = 'index'";
            using var reader = command.ExecuteReader();
            while (reader.Read()) names.Add(reader.GetString(0));
        }

        try { Directory.Delete(directory, recursive: true); } catch (IOException) { }

        Assert.Contains("idx_daily_usage_day", names);
        Assert.Contains("idx_fetch_log_server", names);
    }
}
